### DISCUSSION TITLE
Clubbing .parse and .refine errors/issues in v3

### DISCUSSION THREAD

**Discussion by maneetgoyal** [colinhacks/zod#437] [Q&A] [answered]
We have declared a zod object schema with `.refine` call.

```
const zodSchema = z.object({...}).refine(...); 
```

What is happening in v3 is that only when the criteria laid down by `.object({...})` passes, do we see the `.refine(...)` related errors/issues. In v1,  both classes of errors/issues were shown when we did `zodSchema.parse(...)`. Is there a way to access that behavior again in v3?

**Comment by colinhacks** [comment, marked as answer]
I understand the need for something like this, but it's not likely in the short term. Zod doesn't catch errors that are thrown inside of refinements, so passing an invalid value into a refinement function is likely to throw an unexpected error. Zod schemas *validate types* first, then pass that value through a chain of refinements/transformations. Passing invalid data into this chain is not the intended behavior.

The "right" way to solve this is to differentiate between fatal and non-fatal validation issues. That way if there is a minor validation issue "inside" an object, we could still run the outer refinements. 

```ts
z.object({
  name: z.string().min(5, { nonfatal: true }),
}).refine(val => /* run this even if string is too short */)
```

But this is a major refactor and isn't likely anytime soon. Open to other solutions as well.

**Reply by maneetgoyal** [reply]
Thanks @colinhacks, should I turn this discussion into an issue instead?

**Reply by colinhacks** [reply]
Already open: https://github.com/colinhacks/zod/issues/259
